File: src/main/java/failchat/core/WSClient.java
package failchat.core;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.util.logging.Level;
import java.util.logging.Logger;

public class WSClient {
    private static final Logger logger = Logger.getLogger(WSClient.class.getName());

    private WebSocketClient wsClient;
    private String uri;
    private int reconnectTimeout = 5000;
    ChatClientStatus status = ChatClientStatus.READY;

    public WSClient(String uri) {
        this.uri = uri;
    }

    //non-blocking method
    public void connect() {
        new Thread(() -> {
            while (!(status == ChatClientStatus.SHUTDOWN || status == ChatClientStatus.ERROR)) {
                wsClient = new Wsc(URI.create(uri));
                try {
                    if (wsClient.connectBlocking()) {
                        synchronized (wsClient) {
                            wsClient.wait();
                        }
                    } else {
                        Thread.sleep(reconnectTimeout);
                    }
                } catch (InterruptedException e) {
                    logger.log(Level.WARNING, "Something goes wrong...", e);
                }

            }
        }, "WSReconnectThread").start();
    }

    public void close() {
        status = ChatClientStatus.SHUTDOWN;
        if (wsClient != null) {
            wsClient.close();
        }
        synchronized (wsClient) {
            wsClient.notify();
        }
    }

    public void send(String s) {
        if (status == ChatClientStatus.WORKING) {
            wsClient.send(s);
        }
    }


    public void onOpen(ServerHandshake serverHandshake) {}

    public void onMessage(String s) {}

    public void onClose(int i, String s, boolean b) {}

    public void onReconnect() {}

    public void onError(Exception e) {}

    private class Wsc extends WebSocketClient {
        public Wsc(URI serverURI) {
            super(serverURI);
        }

        @Override
        public void onOpen(ServerHandshake serverHandshake) {
            failchat.core.WSClient.this.status = ChatClientStatus.WORKING;
            failchat.core.WSClient.this.onOpen(serverHandshake);
        }

        @Override
        public void onMessage(String s) {
            failchat.core.WSClient.this.onMessage(s);
        }

        @Override
        public void onClose(int i, String s, boolean b) {
            if (failchat.core.WSClient.this.status == ChatClientStatus.CONNECTING || failchat.core.WSClient.this.status == ChatClientStatus.SHUTDOWN
                    || failchat.core.WSClient.this.status == ChatClientStatus.ERROR) {
                return;
            }
            if (failchat.core.WSClient.this.status == ChatClientStatus.WORKING){
                failchat.core.WSClient.this.status = ChatClientStatus.CONNECTING;
                failchat.core.WSClient.this.onReconnect();
            }
            failchat.core.WSClient.this.onClose(i, s, b);
            synchronized (this) {
                this.notify();
            }
        }

        @Override
        public void onError(Exception e) {
            if (failchat.core.WSClient.this.status == ChatClientStatus.WORKING){
                failchat.core.WSClient.this.status = ChatClientStatus.CONNECTING;
                failchat.core.WSClient.this.onReconnect();
            }
            failchat.core.WSClient.this.onError(e);
        }
    }
}

File: src/main/kotlin/failchat/core/ws/client/WsClient.kt
package failchat.core.ws.client

import failchat.core.ws.client.WsClient.Status.CONNECTING
import failchat.core.ws.client.WsClient.Status.ERROR
import failchat.core.ws.client.WsClient.Status.READY
import failchat.core.ws.client.WsClient.Status.SHUTDOWN
import failchat.core.ws.client.WsClient.Status.WORKING
import failchat.utils.sleep
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.net.URI
import java.time.Duration
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.thread
import kotlin.concurrent.withLock

/**
 * Web socket client with auto-reconnect feature.
 * */
open class WsClient(private val serverUri: URI) {

    private companion object {
        val log: Logger = LoggerFactory.getLogger(WsClient::class.java)
    }

    private val reconnectInterval = Duration.ofSeconds(5)
    private val lock: Lock = ReentrantLock()
    private val reconnectCondition: Condition = lock.newCondition()
    private val status: AtomicReference<Status> = AtomicReference(READY)

    private var wsClient: WebSocketClient = Wsc()

    fun start() {
        thread(start = true, name = "WsClientReconnectThread") {
            tryReconnectLoop()
        }

        log.info("WsClient starter")
    }

    fun stop() {
        status.set(SHUTDOWN)
        lock.withLock { reconnectCondition.signal() }
        wsClient.close()
    }

    fun send(message: String) = wsClient.send(message)

    open fun onOpen(serverHandshake: ServerHandshake) {}

    open fun onMessage(message: String) {}

    open fun onClose(code: Int, reason: String, remote: Boolean) {}

    open fun onReconnect() {}

    open fun onError(e: Exception) {}

    private fun tryReconnectLoop() {
        status.compareAndSet(READY, CONNECTING)

        while (status.get() != SHUTDOWN && status.get() != ERROR) {
            val connected = wsClient.connectBlocking()
            if (!connected) {
                log.warn("Failed to reconnect to {}, next try in {} ms", serverUri, reconnectInterval.toMillis())
                sleep(reconnectInterval)
                wsClient = Wsc()
                continue
            }

            //connected
            log.info("Reconnected to {}", serverUri)
            lock.withLock {
                reconnectCondition.await()
            }

            //prepare to reconnect
            wsClient = Wsc()
        }
        log.info("WsClient stopped")
    }


    private inner class Wsc : WebSocketClient(this@WsClient.serverUri) {

        private val status = this@WsClient.status

        override fun onOpen(serverHandshake: ServerHandshake) {
            status.set(WORKING)
            this@WsClient.onOpen(serverHandshake)
        }

        override fun onMessage(message: String) {
            this@WsClient.onMessage(message)
        }

        override fun onClose(code: Int, reason: String, remote: Boolean) {
            when (status.get()!!) {
                CONNECTING,
                SHUTDOWN,
                ERROR -> return

                WORKING -> {
                    status.set(CONNECTING)
                    this@WsClient.onReconnect()
                }

                READY -> {
                    throw IllegalStateException("$READY")
                }
            }

            this@WsClient.onClose(code, reason, remote)
            lock.withLock { reconnectCondition.signal() }
        }

        override fun onError(e: Exception) {
            val statusChanged = status.compareAndSet(WORKING, CONNECTING)
            if (statusChanged) this@WsClient.onReconnect()

            this@WsClient.onError(e)
        }
    }

    enum class Status {
        READY,
        WORKING,
        CONNECTING,
        ERROR,
        SHUTDOWN
    }

}
